File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

N, M = map(int, input().split())
maps = [list(map(int, list(input().strip()))) for _ in range(N)]

# 이동 방향 (상, 하, 좌, 우)
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

# BFS 함수
def bfs(x, y):
    queue = deque()
    queue.append((x, y))

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            # 범위를 벗어났다면 무시
            if nx < 0 or nx >= N or ny < 0 or ny >= M:
                continue

            # 벽이거나 이미 방문했다면 무시
            if maps[nx][ny] == 0:
                continue

            # 처음 방문하는 곳이면 이전 값 + 1
            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx, ny))

    return maps[N-1][M-1]  # 도착 지점의 값이 곧 최소 칸 수

print(bfs(0, 0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <iostream>
#include <vector>
#include <queue>
#include <string>
using namespace std;

int N, M;
vector<vector<int>> maps;
vector<int> dx = {0, 0, 1, -1};
vector<int> dy = {1, -1, 0, 0};

void bfs(int i, int j) {
    queue<pair<int, int>> q; 
    q.push({i, j});

    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for (int k = 0; k < 4; k++) {
            int nx = x + dx[k];  
            int ny = y + dy[k];

            if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
            if (maps[nx][ny] == 0) continue;

            if (maps[nx][ny] == 1) {
                maps[nx][ny] = maps[x][y] + 1;
                q.push({nx, ny});
            }
        }
    }
}

int main() {
    cin >> N >> M;
    maps.resize(N, vector<int>(M));

    for (int i = 0; i < N; i++) {
        string row;
        cin >> row;
        for (int j = 0; j < M; j++) {
            maps[i][j] = row[j] - '0';  // 문자 '1' → 숫자 1
        }
    }

    bfs(0, 0);

    cout << maps[N - 1][M - 1] << '\n';
    return 0;
}
